infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -283,4 +250,59 @@ public static enum FsmEventType {
         LogReplicationRuntimeEvent,
         LogReplicationSinkEvent
     }
+
+    /**
+     * This class holds the enqueued LogReplicationEvent. Since replication events may run at 0-delay or non-zero delay,
+     * this wrapper class assists in determining if the current event picked by a thread can be processed.
+     * This also ensures that there is only 1 thread active for a session at any time.
+     */
+    private static class ReplicationEventOrderManager{
+        // Contains events with 0-delay. This list helps maintain the order of events processed.
+        private final LinkedList<UUID> processImmediately;
+
+        // This list has the same intention as above, but with a slight difference in kind of event it holds.
+        // This contains the events which have to be scheduled after a non-zero delay.
+        // Currently will have the snapshot-apply-verification events and (only relevant for routing queue model) the check for data function
+        private final LinkedList<UUID> processWithDelay;
+
+        // set to true when an event is being actively processed for the session
+        private boolean currentlyProcessingEvent;
+
+        ReplicationEventOrderManager() {
+            processImmediately = new LinkedList<>();
+            processWithDelay = new LinkedList<>();
+            currentlyProcessingEvent = false;
+        }
+
+        synchronized boolean cannotProcessEvent(UUID eventID) {
+            if (currentlyProcessingEvent) {
+                return true;
+            }
+            boolean cannotProcess = !(!processImmediately.isEmpty() && processImmediately.get(0).equals(eventID) ||
```

Comment:
it can be `canProcess` in this line, but while returning it has be a negation of the value.
This is so we can keep the process methods look similar to each other.

Let me know if I should make the change here...

Reply:
I have implemented approach 1